src/WooCommerce.php

```diff
@@ -1387,10 +1401,15 @@ public static function woocommerce_save_account_details_required_fields(array $f
   /**
    * @implements woocommerce_before_template_part
    */
-  public static function woocommerce_before_template_part($template_name) {
+  public static function woocommerce_before_template_part($template_name, $template_path, $located, $args) {
     if ($template_name === 'myaccount/form-lost-password.php' || $template_name === 'global/form-login.php' || $template_name === 'myaccount/my-subscriptions.php') {
       ob_start();
     }
+    // User input is lost upon any form validation error.
+    // @todo Fix WooCommerce Core to use woocommerce_form_field().
+    if ($template_name === 'myaccount/form-edit-account.php' && !empty($_POST['account_email'])) {
+      $args['user']->user_email = $_POST['account_email'];
```

Comment:
⚠️ The idea to apply this override for the template output variable only is good, but as objects are always passed by reference in PHP, this modification actually changes the global user object and not just the template variable.